### PR TITLE
DDF-3070 Allow the ability to logout if currently guest.

### PR DESCRIPTION
**Pull request by pklinef**
#### What does this PR do?
- Fixed bug that would not allow logging out if guest.
- Fixed IdP logout action provider to use the user’s display name
  instead of their full name id.
- Fixed logout confirmation to URI decode the response message.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@brjeter 
@harrison-tarr 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[Security](https://github.com/orgs/codice/teams/security)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@lessarderic
@stustison

#### How should this be tested? (List steps with links to updated documentation)
Login as guest.  Go to `https://localhost:8993/logout` and verify that you can logout as guest and log in as a different user.

#### What are the relevant tickets?
[DDF-3070](https://codice.atlassian.net/browse/DDF-3070)

#### Screenshots (if appropriate)
idp guest logout
before
after

#### Checklist:
- [ n/a ] Documentation Updated
- [ n/a ] Change Log Updated
- [ n/a ] Update / Add Unit Tests
- [ n/a ] Update / Add Integration Tests

### CODE REVIEW

File: platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java [hunk 1]
```diff
@@ -80,17 +80,14 @@ public Response getActionProviders(@Context HttpServletRequest request)
                             "",
                             true);
 
-                    if (displayName != null
```
Comment (brjeter) [inline platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java]: Did we still want the null check here? In which scenarios would the displayName be null?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java [hunk 2]
```diff
@@ -56,7 +56,9 @@
             try {
 
                 @SuppressWarnings("unchecked")
-                String nameId = SubjectUtils.getName((Subject) ((Map) realmSubjectMap).get("idp"));
+                String nameId = SubjectUtils.getName((Subject) ((Map) realmSubjectMap).get("idp"),
```
Comment (lessarderic) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java]: ❗️ We should add a unit test to cover the case where the name couldn't be found.

File: platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java [hunk 3]
```diff
@@ -80,17 +80,14 @@ public Response getActionProviders(@Context HttpServletRequest request)
...
-                        actionProperties.put("title", action.getTitle());
-                        actionProperties.put("realm", realm);
-                        actionProperties.put("auth", displayName);
-                        actionProperties.put("description", action.getDescription());
-                        actionProperties.put("url",
-                                action.getUrl()
-                                        .toString());
-                        realmToPropMaps.add(actionProperties);
-                    }
+                    actionProperties.put("title", action.getTitle());
```
Comment (lessarderic) [inline platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java]: ❗️ Could we add some unit tests to cover the scenarios where `displayName` is empty and equal to `GUEST_DISPLAY_NAME`?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java [hunk 4]
```diff
@@ -56,7 +56,9 @@
...
+                        "You",
```
Comment (lessarderic) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java]: ❓ Is this `You` string displayed as-is in the UI somehow? If so, is it safe to assume this will get displayed ok in the sentence used by the UI?

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java (hunk 2 above)
Comment (stustison) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java]: SubjectUtils already has unit tests to cover when the name can't be found and it falls back on the default value, which is "You" in this case.

File: platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java (hunk 3 above)
Comment (stustison) [inline platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java]: I think that is already covered by the SubjectUtils unit tests.

File: platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java (hunk 1 above)
Comment (pklinef) [inline platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java]: The second parameter of `SubjectUtils.getName` provides a default value.  Display name should never be null at this point.

File: platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java (hunk 4 above)
Comment (pklinef) [inline platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpLogoutActionProvider.java]: This parameter is the default value for the display name.  In case the current subject does not have a display or name ID then You will be used.  The message gets displayed as "You logged out successfully.".

**Comment by stustison** [comment]
build now

File: platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java (hunk 3 above)
Comment (stustison) [inline platform/security/servlet/security-servlet-logout/src/main/java/org/codice/ddf/security/servlet/logout/LogoutService.java]: I think that we should test that `realmToPropMaps` always has values added to it no matter what the value of the `displayName` is.

**Comment by lessarderic** [comment]
build now

File: platform/security/platform-security-core-api/src/test/java/ddf/security/SubjectUtilsTest.java [hunk 5]
```diff
@@ -112,6 +120,62 @@ public void testGetDefaultName() {
     }
 
     @Test
+    public void testGuestDisplayName() {
+        Subject subject = getSubjectWithPrincipal(new GuestPrincipal("127.0.0.1"));
+        assertEquals(SubjectUtils.GUEST_DISPLAY_NAME, SubjectUtils.getName(subject, null, true));
```
Comment (lessarderic) [inline platform/security/platform-security-core-api/src/test/java/ddf/security/SubjectUtilsTest.java]: ✏️  We could use `assertThat(..., is(...))` instead of `assertEquals()`.

**Comment by pklinef** [comment]
build now
